controllers/flinkcluster_reconciler.go

```diff
@@ -468,31 +468,29 @@ func (reconciler *ClusterReconciler) reconcileJob() (ctrl.Result, error) {
 		var jobID = reconciler.getFlinkJobID()
 		var restartPolicy = observed.cluster.Spec.Job.RestartPolicy
 		var recordedJobStatus = observed.cluster.Status.Components.Job
+		var jobSpec = reconciler.observed.cluster.Spec.Job
 
 		// Update or recover Flink job by restart.
-		var restartJob bool
 		if shouldUpdateJob(observed) {
 			log.Info("Job is about to be restarted to update")
-			restartJob = true
+			err := reconciler.restartJob(*jobSpec.TakeSavepointOnUpgrade)
+			return requeueResult, err
 		} else if shouldRestartJob(restartPolicy, recordedJobStatus) {
 			log.Info("Job is about to be restarted to recover failure")
-			restartJob = true
-		}
-		if restartJob {
-			err := reconciler.restartJob()
-			if err != nil {
-				return requeueResult, err
-			}
-			return requeueResult, nil
+			err := reconciler.restartJob(false)
+			return requeueResult, err
 		}
 
 		// Trigger savepoint if required.
 		if len(jobID) > 0 {
-			if ok, savepointTriggerReason := reconciler.shouldTakeSavepoint(); ok {
-				newSavepointStatus, _ = reconciler.takeSavepointAsync(jobID, savepointTriggerReason)
+			shouldTakeSavepont, savepointTriggerReason := reconciler.shouldTakeSavepoint()
+			if shouldTakeSavepont {
+				err = reconciler.updateSavepointTriggerTimeStatus()
+				if err != nil {
+					newSavepointStatus, _ = reconciler.takeSavepointAsync(jobID, savepointTriggerReason)
+				}
```

Comment:
@shashken @functicons 
It seems that savepoint should be triggered when `err == nil`.
When I tested I have found savepoint was not triggered and only `status.job.lastSavepointTriggerTime` was updated sometimes.

And FlinkCluster status is updated in `updateStatus` function of reconciler, therefore if you have a plan to make a new PR, it might be worth considering how to call the status update function once.

Reply:
Damm I tested the version b4 the CR change with that line, sorry about that